pkg/operator/csi/csidrivercontroller/csi_driver_controller.go

```diff
@@ -0,0 +1,369 @@
+
+	opSpec, opStatus, opResourceVersion, err := c.operatorClient.GetOperatorState()
+	if err != nil {
+		return err
+	}
+
+	// We only support Managed for now
+	if opSpec.ManagementState != opv1.Managed {
+		return nil
+	}
+
+	startTime := time.Now()
+	klog.Info("Starting syncing operator at ", startTime)
+	defer func() {
+		klog.Info("Finished syncing operator at ", time.Since(startTime))
+	}()
+
+	syncErr := c.handleSync(opResourceVersion, meta, opSpec, opStatus)
+	if syncErr != nil {
+		c.eventRecorder.Eventf("SyncError", "Error syncing CSI driver: %s", syncErr)
+	}
+
+	c.updateSyncError(opStatus, syncErr)
+
+	// Update the status using our copy
+	_, _, err = v1helpers.UpdateStatus(c.operatorClient, func(status *opv1.OperatorStatus) error {
+		// Store a copy of our starting conditions, we need to preserve last transition time
+		originalConditions := status.DeepCopy().Conditions
+
+		// Copy over everything else
+		opStatus.DeepCopyInto(status)
+
+		// Restore the starting conditions
+		status.Conditions = originalConditions
+
+		// Manually update the conditions while preserving last transition time
+		for _, condition := range opStatus.Conditions {
+			v1helpers.SetOperatorCondition(&status.Conditions, condition)
+		}
+
+		return nil
+	})
+
+	if err != nil {
+		klog.Errorf("failed to update status: %v", err)
+		if syncErr == nil {
+			syncErr = err
+		}
+	}
+
+	return syncErr
+}
+
+func (c *CSIDriverController) updateSyncError(status *opv1.OperatorStatus, err error) {
+	if err != nil {
+		// Operator is Degraded: could not finish what it was doing
+		v1helpers.SetOperatorCondition(&status.Conditions,
+			opv1.OperatorCondition{
+				Type:    fmt.Sprintf("%s%s", c.name, opv1.OperatorStatusTypeDegraded),
+				Status:  opv1.ConditionTrue,
+				Reason:  "OperatorSync",
+				Message: err.Error(),
+			})
+
+		// Operator is Progressing: some action failed, will try to progress more after exp. backoff.
+		// Do not overwrite existing "Progressing=true" condition to keep its message.
+		cnd := v1helpers.FindOperatorCondition(status.Conditions, opv1.OperatorStatusTypeProgressing)
+		if cnd == nil || cnd.Status == opv1.ConditionFalse {
+			v1helpers.SetOperatorCondition(&status.Conditions,
+				opv1.OperatorCondition{
+					Type:    fmt.Sprintf("%s%s", c.name, opv1.OperatorStatusTypeProgressing),
+					Status:  opv1.ConditionTrue,
+					Reason:  "OperatorSync",
+					Message: err.Error(),
+				})
+		}
+	} else {
+		v1helpers.SetOperatorCondition(&status.Conditions,
+			opv1.OperatorCondition{
+				Type:   fmt.Sprintf("%s%s", c.name, opv1.OperatorStatusTypeDegraded),
+				Status: opv1.ConditionFalse,
+			})
+		// Progressing condition was set in c.handleSync().
+	}
+}
+
+func (c *CSIDriverController) handleSync(
+	resourceVersion string,
+	meta *metav1.ObjectMeta,
+	spec *opv1.OperatorSpec,
+	status *opv1.OperatorStatus,
+) error {
+	var controllerService *appsv1.Deployment
+	if c.controllerManifest != nil {
+		c, err := c.syncDeployment(spec, status)
```

Comment:
my recommendation remains to create separate controllers for these, but with the substitution as it is, we can merge this.

If you had separate controllers, each individual piece becomes an additional controller you run, right?

Reply:
Yes, some CSI operators will need another controller (not from library-go) to fetch some information from the cluster and customize the StorageClass.

I like the idea of having separate controllers, I hadn't considered it before. I haven't addressed that in this PR because it'll require lots of changes (including tests), but I'll work in this next.